docs/benchmarks/profiling.md

```diff
@@ -0,0 +1,122 @@
+[`plutus`](https://github.com/input-output-hk/plutus) project
+[here](https://plutus.readthedocs.io/en/latest/howtos/profiling-scripts.html),
+but points out how this can be done in the `hydra` code base.
+
+## Isolate a transaction to profile
+
+To do any measurements, we need to isolate the actual cardano `Tx` which is to
+be profiled. For example, let's investigate what the `collectCom` transaction
+for `5` parties in the `tx-cost` benchmark is spending most time and memory on.
+
+The benchmark computes many transactions with growing number of participants in `computeCollectComCost`:
+```haskell
+computeCollectComCost =
+  catMaybes <$> mapM compute [1 .. 100]
+ where
+  compute numParties = do
+    (st, tx) <- generate $ genCollectComTx numParties
+    let utxo = getKnownUTxO st
+    case checkSizeAndEvaluate tx utxo of
+      -- [...]
+```
+
+The `tx` here would be the transaction we want to profile, so we can "isolate"
+the transaction for `5` parties by changing the body of this function `maybe []
+pure <$> compute 5`.
+
+## Compiling a script for profiling
+
+The `collectCom` transaction runs `vCommit` and `vHead` validator scripts, so we
+need to add
+
+```
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:profile-all #-}
+```
+
+to the corresponding modules [`Hydra.Contract.Commit`](https://hydra.family/head-protocol/haddock/hydra-plutus/Hydra-Contract-Commit.html) and
+[`Hydra.Contract.Head`](https://hydra.family/head-protocol/haddock/hydra-plutus/Hydra-Contract-Head.html).
+
+## Acquiring an executable script
+
+This can be now achieved using
+[`prepareTxScripts`](https://hydra.family/head-protocol/haddock/hydra-node/Hydra-Ledger-Cardano-Evaluate.html#v:prepareTxScripts).
+We can use this function to acquire and dump the fully applied scripts from the
+transaction onto disk.
+
+```haskell
+-- [...]
+(st, tx) <- generate $ genCollectComTx numParties
+let utxo = getKnownUTxO st
+scripts <- either die pure $ prepareTxScripts tx utxo
+forM_ (zip [1 ..] scripts) $ \(i, s) -> writeFileBS ("scripts-" <> show i <> ".flat") s
+-- [...]
+```
+
+After running the corresponding code (`tx-cost` in our example), we will be left
+with `scripts-{1,2,3,4,5}.flat` files in the current directory.
+
+Unfortunately it's quite hard to tell them apart, but script sizes should help
+in telling the big `vHead` script apart from the smaller `vCommit` script. In
+the profile, names of original `plutus-tx` functions will be retained so that
+should make it clear at the latest.
+
+## Running the script & analysing the results
+
+The tools for this step can be acquired using nix (or alternatively compile the
+upstream projects and use your distribution's package manager):
+
+```
+nix shell nixpkgs#flamegraph github:input-output-hk/plutus#x86_64-linux.plutus.library.plutus-project-924.hsPkgs.plutus-core.components.exes.traceToStacks github:input-output-hk/plutus#x86_64-linux.plutus.library.plutus-project-924.hsPkgs.plutus-core.components.exes.uplc
```

Comment:
question: why `924`? Are we not still using ghc `8.10.7`?

Reply:
Plutus is already supporting `9.2.4` and their nix environment is parameterized by it.